test_conformance/extensions/cl_khr_command_buffer/negative_command_buffer_svm_mem.cpp

```diff
@@ -0,0 +1,413 @@
+//
+// Copyright (c) 2024 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+#include "basic_command_buffer.h"
+#include "svm_command_basic.h"
+#include "procs.h"
+
+
+//--------------------------------------------------------------------------
+namespace {
+
+// CL_INVALID_COMMAND_QUEUE if command_queue is not NULL.
+struct CommandBufferCommandSVMQueueNotNull : public BasicSVMCommandBufferTest
+{
+    using BasicSVMCommandBufferTest::BasicSVMCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_int error = clCommandSVMMemcpyKHR(
+            command_buffer, queue, svm_out_mem(), svm_in_mem(), data_size(), 0,
+            nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(
+            error, CL_INVALID_COMMAND_QUEUE,
+            "clCommandSVMMemcpyKHR should return CL_INVALID_COMMAND_QUEUE",
+            TEST_FAIL);
+
+        error = clCommandSVMMemFillKHR(command_buffer, queue, svm_in_mem(),
+                                       &pattern_1, sizeof(cl_char), data_size(),
+                                       0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(
+            error, CL_INVALID_COMMAND_QUEUE,
+            "clCommandSVMMemFillKHR should return CL_INVALID_COMMAND_QUEUE",
+            TEST_FAIL);
+
+        return CL_SUCCESS;
+    }
+
+    bool Skip() override
+    {
+        return BasicSVMCommandBufferTest::Skip()
+            || is_extension_available(device,
+                                      "cl_khr_command_buffer_multi_device");
+    }
+
+    const cl_char pattern_1 = 0x14;
+};
+
+// CL_INVALID_CONTEXT if the context associated with command_queue,
+// command_buffer, and kernel are not the same.
+struct CommandBufferCommandSVMDifferentContext
+    : public BasicSVMCommandBufferTest
+{
+    using BasicSVMCommandBufferTest::BasicSVMCommandBufferTest;
+
+    cl_int Run() override
+    {
+        cl_int error = clCommandSVMMemcpyKHR(
+            command_buffer, nullptr, svm_out_mem(), svm_in_mem(), data_size(),
+            0, nullptr, nullptr, nullptr);
+
+        test_failure_error_ret(
+            error, CL_INVALID_CONTEXT,
+            "clCommandSVMMemcpyKHR should return CL_INVALID_CONTEXT",
+            TEST_FAIL);
```

Comment:
FWIW, I think it's ambiguous what should happen in this scenario.  The only error condition described in the spec for [clEnqueueSVMMemcpy](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#clEnqueueSVMMemcpy) for CL_INVALID_CONTEXT is:

> [CL_INVALID_CONTEXT](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#CL_INVALID_CONTEXT) if the context associated with command_queue and events in event_wait_list are not the same.

I could certainly imagine this would be an error for some implementations (but which error is not clearly specified).  For other implementations though, such as those that support system SVM, it would probably be more work to generate an error in this scenario than it would be for it to "just work".

Reply:
This test case was removed in the latest commit - works for me.